File: Hash table/hash-table.py
from math import sqrt
from typing import Any

class HashTable:

    __size: int = 0
    __table: list = []

    def __isprime(self, n: int) -> bool:
        result: list = []
        d: int = 2
        while d * d <= n:
            if n % d == 0:
                result.append(d)
                n //= d
            else:
                d += 1
        if n > 1:
            result.append(n)
        if len(result) > 1: return False
        else: return True

    def __nearest_prime(self, n: int) -> int:
        if n % 2 == 0:
            n += 1
        while not self.__isprime(n):
            n += 2
        return n

    def __count_hash(self, s: str) -> int:
        n: int = 7
        for c in s:
            n = n * 31 + ord(c)
        return int(self.__size * (n * ((sqrt(5) - 1) / 2) % 1))

    def __init__(self, size: int) -> None:
        self.__size = self.__nearest_prime(size)
        self.__table = [None for _ in range(0, self.__size)]

    def search(self, key: str) -> Any:
        return self.__table[self.__count_hash(key)]

    def insert(self, key: str, value: Any) -> None:
        self.__table[self.__count_hash(key)] = value

    def remove(self, key: str) -> None:
        self.__table[self.__count_hash(key)] = None

File: Hash table/hash-table.go
package main

import (
	"math"
)

func IsPrime(n int) bool {
	result := []int{}
	d := 2
	for d*d <= n {
		if n%d == 0 {
			result = append(result, d)
			n = int(n / d)
		} else {
			d++
		}
	}
	if n > 1 {
		result = append(result, n)
	}
	if len(result) > 1 {
		return false
	} else {
		return true
	}
}

func NearestPrime(n int) int {
	if n%2 == 0 {
		n += 1
	}
	for !IsPrime(n) {
		n += 2
	}
	return n
}

type HashTable[T any] struct {
	Size  int
	Table []T
}

func NewHashTable[T any](size int) HashTable[T] {
	sz := NearestPrime(size)
	table := []T{}
	var def T
	for i := 0; i < size; i++ {
		table = append(table, def)
	}
	ht := HashTable[T]{sz, table}
	return ht
}

func (ht HashTable[T]) CountHash(s string) int {
	n := 7
	for i := 0; i < len(s); i++ {
		n = n*31 + int(s[i])
	}
	return int(float64(ht.Size) * (float64(n)*((math.Sqrt(5)-1)/2) - math.Floor(float64(n)*((math.Sqrt(5)-1)/2))))
}

func (ht HashTable[T]) Search(key string) T {
	return ht.Table[ht.CountHash(key)]
}

func (ht HashTable[T]) Insert(key string, value T) {
	ht.Table[ht.CountHash(key)] = value
}

func (ht HashTable[T]) Remove(key string) {
	var def T
	ht.Table[ht.CountHash(key)] = def
}

func main() {
}
